codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -408,13 +442,17 @@ private void applyLocalSettings(Http2Settings settings) throws Http2Exception {
         }
 
         @Override
-        public void onSettingsRead(ChannelHandlerContext ctx, Http2Settings settings) throws Http2Exception {
-            // Acknowledge receipt of the settings. We should do this before we process the settings to ensure our
-            // remote peer applies these settings before any subsequent frames that we may send which depend upon these
-            // new settings. See https://github.com/netty/netty/issues/6520.
-            encoder.writeSettingsAck(ctx, ctx.newPromise());
-
-            encoder.remoteSettings(settings);
+        public void onSettingsRead(final ChannelHandlerContext ctx, Http2Settings settings) throws Http2Exception {
+            if (outstandingRemoteSettingsQueue == null) {
+                // Acknowledge receipt of the settings. We should do this before we process the settings to ensure our
+                // remote peer applies these settings before any subsequent frames that we may send which depend upon
+                // these new settings. See https://github.com/netty/netty/issues/6520.
+                encoder.writeSettingsAck(ctx, ctx.newPromise());
+
+                encoder.remoteSettings(settings);
+            } else {
+                outstandingRemoteSettingsQueue.add(settings);
```

Comment:
@Scottmitch did you consider having not a queue but a callback that may return a `Future` that will be notified when the ack should be written ? From a quick thought this seems "cleaner".

Reply:
The way the h2 protocol works the settings ACKs apply the earliest non-acked settings frame in a sequential fashion. So if this is future based, or if we rely upon the user to provide us the settings in the `SettingsAckFrame` then we may allow for applying settings locally out of order and become out of sync with our remote peer. I think we would need some "queue" (either chaining futures or a queue like as here) in order to preserve the correct ordering. Did you have something else in mind?

Reply:
yeah I was thinking about chaining `Future`. Not sure if this is better but a `Queue` just felt a bit strange as you can do all kind of "strange" things with a `Queue`

Reply:
I think a `Queue` is a more straightforward approach, because the encoder needs a settings synchronously when there is an ACK written. If you don't agree can you mock up what you had in mind from the producer/consumer perspective so I can understand better?

We could expose it as something like a `Supplier<Http2Settings>` (which we would have to make our own supplier bcz of jdk6 compatibility) to avoid the over exposure. Do we have something similar in Netty yet, and do you feel it is worth adding this type of API?

FWIW `Http2ConnectionEncoder` has a `Http2Settings pollSentSettings()` for the same reason (so the decoder can get the last sent settings when it reads an ack) if we could break APIs I would do something similar here.

Reply:
hmm ok... Then lets just use a `Queue` and keep it simple. We need to revisit all of this for 5.0 anyway.